module/optee/voltd_regulator/src/mod_optee_voltd_regulator.c

```diff
@@ -0,0 +1,366 @@
+
+        if (desc->type == VOLTAGE_TYPE_FULL_LIST) {
+            info->name = regulator_name(ctx->regulator);
+            info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+            info->level_range.level_count = desc->num_levels;
+            info->level_range.min_uv = desc->entries[0];
+            info->level_range.max_uv = desc->entries[desc->num_levels - 1];
+        } else if (desc->type == VOLTAGE_TYPE_INCREMENT) {
+            info->name = regulator_name(ctx->regulator);
+            info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+            info->level_range.min_uv = desc->entries[0];
+            info->level_range.max_uv = desc->entries[1];
+            info->level_range.step_uv = desc->entries[2];
+        } else {
+            fwk_assert(0);
+            return FWK_E_PANIC;
+        }
+    }
+
+    FWK_LOG_DEBUG(
+        MOD_PREFIX "Get info for %u/%s: range [%d %d]",
+        fwk_id_get_element_idx(dev_id), info->name,
+        info->level_range.min_uv,
+        info->level_range.max_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int optee_voltd_regulator_level_from_index(fwk_id_t dev_id,
+                                                  unsigned int index,
+                                                  int32_t *level_uv)
+{
+    struct optee_voltd_regulator_dev_ctx *ctx;
+    struct regulator_voltages *desc;
+
+    ctx = find_ctx(dev_id);
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (ctx->regulator == NULL) {
+        /* Treat unexposed voltage domain a stubbed 0V fixed level regulator */
+        if (index > 0) {
+            return FWK_E_RANGE;
+        }
+
+        *level_uv = 0;
+
+        return FWK_SUCCESS;
+    }
+
+    if (regulator_supported_voltages(ctx->regulator, &desc) != TEE_SUCCESS) {
+        return FWK_E_SUPPORT;
+    }
+
+    if (desc->type != VOLTAGE_TYPE_FULL_LIST) {
+        return FWK_E_DATA;
+    }
+    if (index >= desc->num_levels) {
+        return FWK_E_RANGE;
+    }
+
+    *level_uv = (int32_t)desc->entries[index];
+
+    FWK_LOG_DEBUG(
+        MOD_PREFIX "Get level from index for %u/%s: index %u, level %"PRId32"uV",
+        fwk_id_get_element_idx(dev_id), regulator_name(ctx->regulator),
+        index, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static const struct mod_voltd_drv_api api_optee_regu = {
+    .get_level = optee_voltd_regulator_get_level,
+    .set_level = optee_voltd_regulator_set_level,
+    .set_config = optee_voltd_regulator_set_config,
+    .get_config = optee_voltd_regulator_get_config,
+    .get_info = optee_voltd_regulator_get_info,
+    .get_level_from_index = optee_voltd_regulator_level_from_index,
+};
+
+/*
+ * Framework handler functions
+ */
+
+static int optee_voltd_regulator_init(fwk_id_t module_id,
+                                      unsigned int element_count,
+                                      const void *data)
+{
+    module_ctx.dev_count = element_count;
+
+    if (element_count > 0) {
```

Comment:
is it allowed to have 0 elements?

Reply:
It would be more flexible we can enable voltage domain config switches and let the platform tell whether or not some OP-TEE regulators are exposed as voltage domains. Looking at module voltage_domain, i see the init function begins with:
```c
    if (element_count == 0)
        return FWK_SUCCESS;
```

Reply:
OK, if `0` elements is an allowed API implementation should be revisited, only `elements`  are allowed, not `module` as id

Reply:
Ok, I think it's ok on my side to require at least 1 element when this module is embedded. I'll fix.